File: engine/src/service.ts
import { Schema } from "effect";

export const ServiceSchema = Schema.Struct({
  checkedAt: Schema.NullOr(Schema.String),
  localUrl: Schema.String,
  name: Schema.String,
  originStatus: Schema.NullOr(Schema.Literals(["online", "offline"])),
  publicStatus: Schema.Literals(["reachable", "unknown", "unreachable"]),
  publicUrl: Schema.NullOr(Schema.String),
  shared: Schema.Boolean,
  tunnelStatus: Schema.NullOr(
    Schema.Literals(["connected", "connecting", "error"])
  ),
});

export type Service = Schema.Schema.Type<typeof ServiceSchema>;

export const registeredService = (name: string, localUrl: string): Service => ({
  checkedAt: null,
  localUrl,
  name,
  originStatus: null,
  publicStatus: "unknown",
  publicUrl: null,
  shared: false,
  tunnelStatus: null,
});

File: engine/src/service.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddService {
    pub name: String,
    pub local_url: String,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Service {
    pub name: String,
    pub shared: bool,
    pub origin_status: Option<OriginStatus>,
    pub tunnel_status: Option<TunnelStatus>,
    pub public_status: PublicStatus,
    pub local_url: String,
    pub public_url: Option<String>,
    pub checked_at: Option<String>,
}

impl Service {
    pub fn registered(name: String, local_url: String) -> Self {
        Self {
            name,
            shared: false,
            origin_status: None,
            tunnel_status: None,
            public_status: PublicStatus::Unknown,
            local_url,
            public_url: None,
            checked_at: None,
        }
    }
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum OriginStatus {
    Online,
    Offline,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum TunnelStatus {
    Connected,
    Connecting,
    Error,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum PublicStatus {
    Unknown,
    Reachable,
    Unreachable,
}
